invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,193 @@
+func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+	return &InvoiceExpiryWatcher{
+		clock:       clock,
+		newInvoices: make(chan *invoiceExpiry),
+		quit:        make(chan struct{}),
+	}
+}
+
+// Start starts the the subscription handler and the main loop. Start() will
+// return with error if InvoiceExpiryWatcher is already started. Start()
+// expects a cancellation function passed that will be use to cancel expired
+// invoices by their payment hash.
+func (ew *InvoiceExpiryWatcher) Start(
+	cancelInvoice func(lntypes.Hash) error) error {
+
+	ew.Lock()
+	defer ew.Unlock()
+
+	if ew.started {
+		return fmt.Errorf("InvoiceExpiryWatcher already started")
+	}
+
+	ew.started = true
+	ew.cancelInvoice = cancelInvoice
+	ew.wg.Add(1)
+	go ew.mainLoop()
+
+	return nil
+}
+
+// Stop quits the expiry handler loop and waits for InvoiceExpiryWatcher to
+// fully stop.
+func (ew *InvoiceExpiryWatcher) Stop() {
+	ew.Lock()
+	defer ew.Unlock()
+
+	if ew.started {
+		// Signal subscriptionHandler to quit and wait for it to return.
+		close(ew.quit)
+		ew.wg.Wait()
+		ew.started = false
+	}
+}
+
+// AddInvoice adds a new invoice to the InvoiceExpiryWatcher. This won't check
+// if the invoice is already added and will only add invoices with ContractOpen
+// state.
+func (ew *InvoiceExpiryWatcher) AddInvoice(
+	paymentHash lntypes.Hash, invoice *channeldb.Invoice) {
+
+	if invoice.State != channeldb.ContractOpen {
+		log.Debugf("Invoice not added to expiry watcher: %v", invoice)
+		return
+	}
+
+	realExpiry := invoice.Terms.Expiry
+	if realExpiry == 0 {
+		realExpiry = zpay32.DefaultInvoiceExpiry
+	}
+
+	expiry := invoice.CreationDate.Add(realExpiry)
+
+	log.Debugf("Adding invoice '%v' to expiry watcher, expiration: %v",
+		paymentHash, expiry)
+
+	newInvoiceExpiry := &invoiceExpiry{
+		PaymentHash: paymentHash,
+		Expiry:      expiry,
+	}
+
+	select {
+	case ew.newInvoices <- newInvoiceExpiry:
+	case <-ew.quit:
+		// Select on quit too so that callers won't get blocked in case
+		// of concurrent shutdown.
+	}
+}
+
+// nextExpiry returns a Time chan to wait on until the next invoice expires.
+// If there are no active invoices, then it'll simply wait indefinitely.
+func (ew *InvoiceExpiryWatcher) nextExpiry() <-chan time.Time {
+	if !ew.expiryQueue.Empty() {
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		return time.After(top.Expiry.Sub(ew.clock.Now()))
+	}
+
+	return nil
+}
+
+// cancelExpiredInvoices will cancel all expired invoices and removes them from
+// the expiry queue.
+func (ew *InvoiceExpiryWatcher) cancelExpiredInvoices() {
+	for !ew.expiryQueue.Empty() {
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		if !top.Expiry.Before(ew.clock.Now()) {
+			break
+		}
+
+		log.Infof("Cancelling expired invoice '%v', expiry: %v",
+			top.PaymentHash, top.Expiry)
+
+		err := ew.cancelInvoice(top.PaymentHash)
+		if err != nil && err != channeldb.ErrInvoiceAlreadySettled &&
```

Comment:
Does this potentially cancel an invoice in the `Accepted` state? That shouldn't happen.

Reply:
Thanks for pointing this out. As discussed offline, this is now handled in `InvoiceRegistry.cancelInvoiceImpl`.